repo/blob/pit/pit.go

```diff
@@ -0,0 +1,161 @@
+// Package pit provides a Point-in-time view of a versioned blob store
+package pit
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/blob/readonly"
+)
+
+type pitStorage struct {
+	blob.Storage
+
+	vs          VersionedStorage
+	pointInTime time.Time
+}
+
+func (s pitStorage) ListBlobs(ctx context.Context, blobIDPrefix blob.ID, cb func(bm blob.Metadata) error) error {
+	var (
+		previousID blob.ID
+		vs         []VersionMetadata
+	)
+
+	err := s.vs.ListBlobVersions(ctx, blobIDPrefix, func(vm VersionMetadata) error {
+		if vm.BlobID != previousID {
+			// different blob, process previous one
+			if v, found := newestAtUnlessDeleted(vs, s.pointInTime); found {
+				if err := cb(v.Metadata); err != nil {
+					return err
+				}
+			}
+
+			previousID = vm.BlobID
+			vs = vs[:0] // reset for next blob
+		}
+
+		vs = append(vs, vm)
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "could not list blob versions at time %s", s.pointInTime)
+	}
+
+	// process last blob
+	if v, found := newestAtUnlessDeleted(vs, s.pointInTime); found {
+		if err := cb(v.Metadata); err != nil {
+			return err
```

Comment:
Just an observation that if the callback throws an error here, it'll behave differently, i.e. have a reduced wrapped-context, compared to if it's thrown in the `ListBlobVersions` call above. Seems ideally the error behavior should be identical whether the callback hits an error on the last blob or not.

From reading the code, looks like a callback error would wrap `"callback failed for <key>"` and `"could not list blob versions at time <point-in-time>"`, unless it was called here on the last blob. An error in the last blob's CB won't include that extra info.